lib/concurrent/synchronization/immutable_struct.rb

```diff
@@ -0,0 +1,67 @@
+module Concurrent
+  module Synchronization
+    # Similar to Struct but the fields are immutable and always visible (like Java final fields)
+    # @example
+    #   Person = ImmutableStruct.with_fields :name, :age
+    #   Person.new 'John Doe', 15
+    #   Person['John Doe', 15]
+    #   Person['John Doe', 15].members # => [:name, :age]
+    #   Person['John Doe', 15].values  # => ['John Doe', 15]
+    class ImmutableStruct < Synchronization::Object
+      def self.with_fields(*names, &block)
+        Class.new(self) do
+          attr_reader(*names)
+
+          class_eval <<-RUBY, __FILE__, __LINE__ + 1
```

Comment:
Only performance improvements in some cases, otherwise define_method is cleaner. I wanted to avoid storing instance variables using reflection (which is optimized only on Truffle AFAIK).